sklearn/neighbors/lmnn.py

```diff
@@ -0,0 +1,853 @@
+                                 'dimensionality {}!'.format(
+                                    self.n_features_out, len(X[0])))
+
+        # Check preferred number of neighbors
+        max_neighbors = min_class_size - 1
+        if self.n_neighbors > max_neighbors:
+            warnings.warn('n_neighbors(={}) too high. Setting to {}.'.
+                          format(self.n_neighbors, max_neighbors))
+        self.n_neighbors_ = min(self.n_neighbors, max_neighbors)
+        # TODO: Notify superclass KNeighborsClassifier that n_neighbors
+        # might have changed to n_neighbors_
+        # super(LargeMarginNearestNeighbor, self).set_params(
+        # n_neighbors=self.n_neighbors_)
+
+        return X, y_inversed
+
+    def _init_transformer(self, X):
+        """Initialize the linear transformation by setting to user specified
+        parameter, loading from a file, applying PCA or setting to identity.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features_in)
+            Data samples.
+
+        Returns
+        -------
+        L : array, shape (n_features_out, n_features_in)
+            The initial linear transformation.
+
+        """
+
+        if self.L is not None:
+            L = np.asarray(self.L)
+        elif self.warm_start:
+            L = self.L_
+        elif self.use_pca and X.shape[1] > 1:
+            cov_ = np.cov(X, rowvar=False)  # Mean is removed
+            _, evecs = np.linalg.eigh(cov_)
+            evecs = np.fliplr(evecs)  # Sort by descending eigenvalues
+            L = evecs.T  # Get as eigenvectors as rows
+        else:
+            L = np.eye(X.shape[1])
+
+        if self.n_features_out is None:
+            n_features_out = L.shape[0]
+        else:
+            n_features_out = self.n_features_out
+            if L.shape[0] > n_features_out:
+                warnings.warn('Decreasing the initial linear transformation '
+                              'output dimensionality ({}) to the '
+                              'preferred output dimensionality ({}).'.
+                              format(L.shape[0], n_features_out),
+                              DataDimensionalityWarning)
+                L = L[:n_features_out]
+
+        return L, n_features_out
+
+    def _select_target_neighbors(self, X, y):
+        """Find the target neighbors of each sample, that stay fixed during
+        training.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features_in)
+            The training samples.
+
+        y : array, shape (n_samples,)
+            The corresponding training labels indices.
+
+        Returns
+        -------
+        target_neighbors: array, shape (n_samples, n_neighbors)
+            An array of neighbors indices for each sample.
+
+        """
+
+        target_neighbors = np.empty((X.shape[0], self.n_neighbors_), dtype=int)
+        for class_num in range(len(self.classes_)):
```

Comment:
I would still have to call `kneighbors()` on the samples of each class, which is essentially what I do. I just copied the needed lines from there. So I guess this is faster, since I avoid creating any new objects, right?